File: docs/ASL/MapGenerators/Hex Generators/src/lib/terrain-style.js
// lib/terrain-style.js
// Presentation tokens & helpers for ASL terrain rendering.
// - Distinguishes BASE terrains (ground layer) vs OVERLAYS (on top).
// - Centralizes base underpaint colors and pattern-id selection.
// - Exposes helpers used by renderers and the legend.

///////////////////////////////
// Canonical terrain classes //
///////////////////////////////

// Ground layer (what the hex "is"). These ids are used throughout.
export const BASE_TERRAINS = new Set([
  'open', 'woods', 'orchard', 'brush', 'grain', 'marsh', 'sand', 'scrub',
]);

// Overlay categories (sit on top of the base). More can be added later.
export const OVERLAY_TYPES = {
  // Pattern ids for buildings come from the "viz" <defs>:
  //   'wood', 'stone1' (1 level), 'stone2' (2 levels)
  building: new Set(['wood', 'stone1', 'stone2']),
};

// Convenience checks
export const isBaseTerrain   = (t) => BASE_TERRAINS.has(t);
export const isBuildingLayer = (id) => OVERLAY_TYPES.building.has(id);

///////////////////////////////
// Colors (underpaint fill)  //
///////////////////////////////

export const COLORS = {
  openBase:  '#90a955', // olive
  woodsBase: '#6f8f3d',
  orchardBase: '#7da35f',
  brushBase: '#b6c38a',
  grainBase: '#d9c178', // <-- ensure Grain is golden everywhere
  marshBase: '#9db9a4',
  sandBase:  '#e3cf9e',
  scrubBase: '#9fb389',

  labelDark:  '#111',
  labelLight: '#ffffff',
};

// Solid fill color under the pattern for a given base terrain.
export function baseFillColor(base) {
  switch (base) {
    case 'woods':   return COLORS.woodsBase;
    case 'orchard': return COLORS.orchardBase;
    case 'brush':   return COLORS.brushBase;
    case 'grain':   return COLORS.grainBase;   // key for Grain parity
    case 'marsh':   return COLORS.marshBase;
    case 'sand':    return COLORS.sandBase;
    case 'scrub':   return COLORS.scrubBase;
    default:        return COLORS.openBase;    // 'open' and unknowns
  }
}

// Label color chosen for contrast against the base fill.
export function labelColor(base) {
  // Woods and darker greens benefit from light labels; others go dark.
  switch (base) {
    case 'woods':
      return COLORS.labelLight;
    default:
      return COLORS.labelDark;
  }
}

///////////////////////////////
// Pattern id selection      //
///////////////////////////////

// v39 base terrain patterns (must match ids defined in terrain-defs.js)
const PATTERN_BY_BASE = {
  open:    'openGroundPattern',
  woods:   'woodsPattern',
  orchard: 'orchardPattern',
  brush:   'brushPattern',
  grain:   'grainPattern',
  marsh:   'marshPattern',
  sand:    'sandPattern',
  scrub:   'scrubPattern',
};

export function patternIdForBase(base) {
  return PATTERN_BY_BASE[base] || PATTERN_BY_BASE.open;
}

// Building overlay → viz pattern id.
// Accepts either a string id ('wood'|'stone1'|'stone2')
// or an object like { type:'stone'|'wood', levels:1|2 }.
export function patternIdForBuilding(building) {
  if (!building) return null;

  if (typeof building === 'string') {
    const id = building.toLowerCase();
    return OVERLAY_TYPES.building.has(id) ? id : null;
  }

  const type = (building.type || '').toLowerCase();
  const lvl  = Number(building.levels || building.level || 1);

  if (type === 'wood') return 'wood';
  if (type === 'stone') return lvl >= 2 ? 'stone2' : 'stone1';

  return null;
}

///////////////////////////////
// Normalization & tracking  //
///////////////////////////////

// Convert a template/base string into a canonical base id in BASE_TERRAINS.
export function normalizeBase(templateOrString) {
  const raw = typeof templateOrString === 'string'
    ? templateOrString
    : (templateOrString?.baseTerrain || templateOrString?.base || '');

  const s = String(raw).trim().toLowerCase();

  // Common aliases from UI/JSON
  if (s === 'openground' || s === 'open ground' || s === 'open') return 'open';
  if (s === 'light woods' || s === 'lightwoods' || s === 'woods') return 'woods';
  if (s === 'orchard') return 'orchard';
  if (s === 'brush')   return 'brush';
  if (s === 'grain')   return 'grain';
  if (s === 'marsh')   return 'marsh';
  if (s === 'sand')    return 'sand';
  if (s === 'scrub')   return 'scrub';

  // Fallback to open so we always render something
  return 'open';
}

/**
 * Track which base/overlay patterns were used for legend construction.
 * Mutates the provided `used` set object in-place:
 *   used = { bases: Set<string>, buildings: Set<string> }
 */
export function trackUsage(used, template) {
  if (!used) return;

  const base = normalizeBase(template);
  used.bases?.add(base);

  const bpid = patternIdForBuilding(template?.building);
  if (bpid) used.buildings?.add(bpid);
}

File: src/Legacy/Samples/ASL/AslHexMap.Core/Rendering/TerrainStyle.cs
using AslHexMap.Core.Schema;

namespace AslHexMap.Core.Rendering
{
    /// <summary>
    /// Presentation tokens & helpers for ASL terrain rendering.
    /// - Distinguishes BASE terrains (ground layer) vs overlays (buildings).
    /// - Centralizes base underpaint colors and pattern-id selection.
    /// - Handles common synonyms from JSON (e.g., "openGround" → "open").
    /// </summary>
    public static class TerrainStyle
    {
        /////////////////////////////////////////
        // Canonical terrain ids (ground layer) //
        /////////////////////////////////////////

        // Ground/base terrains we currently support.
        public static readonly HashSet<string> BaseTerrains = new(StringComparer.OrdinalIgnoreCase)
        {
            "open", "woods", "orchard", "brush", "grain", "marsh", "sand", "scrub"
        };

        // Overlay categories (drawn above the base).
        public static class OverlayTypes
        {
            // Building “materials” supported by our defs.
            public static readonly HashSet<string> Building = new(StringComparer.OrdinalIgnoreCase)
            {
                "stone","stone1","stone2","wood"
            };
        }

        /////////////////////////////////////////
        // Synonyms / normalization            //
        /////////////////////////////////////////

        // Map common inputs to canonical ids (after NormalizeKey).
        private static readonly Dictionary<string, string> Synonyms = new(StringComparer.OrdinalIgnoreCase)
        {
            // Open Ground variants
            ["openground"] = "open",
            ["open-ground"] = "open",
            ["open_ground"] = "open",
            ["og"] = "open",

            // Add more synonyms here if needed:
            // ["lightwoods"] = "woods",
            // ["orchards"]   = "orchard",
        };

        private static string NormalizeKey(string s)
            => s.Replace(" ", "").Replace("_", "").Replace("-", "").ToLowerInvariant();

        /// <summary>
        /// Normalize any template/base input to a canonical base terrain id. Defaults to "open".
        /// </summary>
        public static string NormalizeBase(string? baseTerrain)
        {
            if (string.IsNullOrWhiteSpace(baseTerrain)) return "open";
            var keyNorm = NormalizeKey(baseTerrain.Trim());

            if (Synonyms.TryGetValue(keyNorm, out var mapped))
                return mapped;

            // If already a canonical id (case-insensitive), return it lowercased.
            return BaseTerrains.Contains(keyNorm) ? keyNorm : "open";
        }

        /////////////////////////////////////////
        // Pattern id selection                //
        /////////////////////////////////////////

        /// <summary>
        /// Return the pattern id to use for a base terrain (for fill="url(#id)").
        /// </summary>
        public static string PatternIdForBase(string? baseTerrain)
        {
            return NormalizeBase(baseTerrain) switch
            {
                "open" => "openPattern",
                "woods" => "woodsPattern",
                "orchard" => "orchardPattern",
                "brush" => "brushPattern",
                "grain" => "grainPattern",
                "marsh" => "marshPattern",
                "sand" => "sandPattern",
                "scrub" => "scrubPattern",
                _ => "openPattern"
            };
        }

        /// <summary>
        /// Return the pattern id for a building overlay given its material/type.
        /// Expected values: stone1, stone2, wood. Returns null if unknown.
        /// </summary>
        public static string? PatternIdForBuilding(string? buildingMaterial)
        {
            if (string.IsNullOrWhiteSpace(buildingMaterial)) return null;
            var key = buildingMaterial.Trim().ToLowerInvariant();
            return OverlayTypes.Building.Contains(key) ? key : null; // ids match defs: "stone1"/"stone2"/"wood"
        }

        public static string? PatternIdForBuilding(BuildingSpec? spec)
        {
            if (spec is null) return null;
            var t = (spec.Type ?? "").Trim().ToLowerInvariant();

            // normalize "wooden" → "wood"
            if (t == "wooden") t = "wood";

            if (t == "wood") return "wood";

            if (t == "stone")
            {
                var levels = spec.Levels.GetValueOrDefault(1);
                return levels >= 2 ? "stone2" : "stone1";
            }

            // future: "rubble", "factory", etc.
            return null;
        }


    /////////////////////////////////////////
    // Colors (underpaint/fallback fills)  //
    /////////////////////////////////////////

    public static class Colors
        {
            public const string OpenBase = "#90a955"; // olive
            public const string WoodsBase = "#6f8f3d";
            public const string OrchardBase = "#7da35f";
            public const string BrushBase = "#8aa15f";
            public const string GrainBase = "#d5c07a";
            public const string MarshBase = "#7aa08a";
            public const string SandBase = "#d8c9a6";
            public const string ScrubBase = "#9aa76f";

            /// <summary>Fallback/underpaint color for a base terrain.</summary>
            public static string ForBase(string baseTerrain) => TerrainStyle.NormalizeBase(baseTerrain) switch
            {
                "open" => OpenBase,
                "woods" => WoodsBase,
                "orchard" => OrchardBase,
                "brush" => BrushBase,
                "grain" => GrainBase,
                "marsh" => MarshBase,
                "sand" => SandBase,
                "scrub" => ScrubBase,
                _ => OpenBase
            };
        }

        /////////////////////////////////////////
        // Legend usage tracking (optional)    //
        /////////////////////////////////////////

        /// <summary>
        /// Track which base/overlay patterns were used, to build a legend later.
        /// Mutates the provided sets in-place.
        /// </summary>
        public static void TrackUsage(
            (ISet<string> Bases, ISet<string> Buildings) used,
            string? baseTerrain,
            string? buildingMaterial)
        {
            used.Bases.Add(NormalizeBase(baseTerrain));
            var bpid = PatternIdForBuilding(buildingMaterial);
            if (bpid is not null) used.Buildings.Add(bpid);
        }
    }
}
